packages/cicero-core/src/template.js

```diff
@@ -172,15 +178,60 @@ class Template {
         return hasher.digest('hex');
     }
 
+    /**
+     * signs a string made up of template hash and time stamp using private key derived
+     * from the keystore
+     * @param {String} [p12File] - encoded string of p12 keystore file
+     * @param {String} [passphrase] - passphrase for the keystore file
+     * @param {Number} [timeStamp] - timeStamp of the moment of signature is done
+     * @private
+     */
+    signTemplate(p12File, passphrase, timeStamp) {
+        const templateHash = this.getHash();
+        // decode p12 from base64
+        const p12Der = forge.util.decode64(p12File);
+        // get p12 as ASN.1 object
+        const p12Asn1 = forge.asn1.fromDer(p12Der);
+        // decrypt p12 using the passphrase 'password'
+        const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, false, passphrase);
+        //X509 cert forge type
+        const certificateForge = p12.safeContents[0].safeBags[0].cert;
+        //Private Key forge type
+        const privateKeyForge = p12.safeContents[1].safeBags[0].key;
+        //convert cert and private key from forge to PEM
+        const certificatePem = forge.pki.certificateToPem(certificateForge);
+        const privateKeyPem = forge.pki.privateKeyToPem(privateKeyForge);
+        //convert private key in pem to private key type in node
+        const privateKey = crypto.createPrivateKey(privateKeyPem);
+        const sign = crypto.createSign('SHA256');
+        sign.write(templateHash + timeStamp);
+        sign.end();
+        const signature = sign.sign(privateKey, 'hex');
+        const signatureObject = {
+            templateHash: templateHash,
+            timeStamp: timeStamp,
+            signatoryCert: certificatePem,
+            signature: signature
+        };
+        this.authorSignature = signatureObject;
+    }
+
     /**
      * Persists this template to a Cicero Template Archive (cta) file.
      * @param {string} [language] - target language for the archive (should be 'ergo')
-     * @param {Object} [options] - JSZip options
-     * @param {Buffer} logo - Bytes data of the PNG file
+     * @param {Object} [options] - JSZip options and keystore object containing path and passphrase for the keystore
      * @return {Promise<Buffer>} the zlib buffer
      */
     async toArchive(language, options) {
-        return TemplateSaver.toArchive(this, language, options);
+        if (!this.authorSignature) {
+            if (options.keyStore) {
+                const timeStamp = Date.now();
+                this.signTemplate(options.keyStore.p12File, options.keyStore.passphrase, timeStamp);
+            }
+            return TemplateSaver.toArchive(this, language, options);
+        } else {
+            throw new Error('Template is already signed by the author.');
```

Comment:
Can you help me to understand this, please?

This means that once I load a signed template with the JS API, I can't generate the archive again, right? Which means that I can't make derivative templates for a signed template?

Reply:
Yes you are correct.

I actually wanted to add some restrictions on signing template again because of the following problem.

If a `developer A `has authored a template and published it on template studio. A `developer B` should not be able to sign that same template and try to publish it. I figured it out that this is not really a good approach to solve this. I wanted to discuss about the same.

Reply:
@jeromesimeon @martinhalford any thoughts?

My worry is that this is would prevent some intended behaviours. For example, say that you wanted to extend an AP template (which it's license allows), but AP signed the original template so you can't extend it through any tooling.

Also, this doesn't technically stop you creating a derivative work, it just makes it a bit more difficult (you have to break open the archive and remove the signature.json) file.

I think that what we really want is to stop you creating a derivative work that you then pass off as coming from someone else. Beyond that, the license restricts what you're legally allowed to do with the code.